test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1277 @@
+			    != (int)pktio->num_tx_qs) {
+				ODPH_ERR("Error querying packet I/O output queue (%s)\n",
+					 pktio->name);
+				return false;
+			}
+		}
+
+		odp_pktio_config_init(&pktio_config);
+
+		if (odp_pktio_config(pktio->handle, &pktio_config) < 0) {
+			ODPH_ERR("Error configuring packet I/O extra options (%s)\n", pktio->name);
+			return false;
+		}
+
+		(void)odp_pktio_mac_addr(pktio->handle, &pktio->src_mac, sizeof(pktio->src_mac));
+
+		if (odp_pktio_start(pktio->handle) < 0) {
+			ODPH_ERR("Error starting packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static odp_bool_t setup_fwd_table(prog_config_t *config)
+{
+	fwd_entry_t *fwd_e;
+
+	config->fwd_tbl = odph_iplookup_table_create(PROG_NAME, 0U, 0U, sizeof(fwd_entry_t *));
+
+	if (config->fwd_tbl == NULL) {
+		ODPH_ERR("Error creating forwarding table\n");
+		return false;
+	}
+
+	for (uint32_t i = 0U; i < config->num_fwds; ++i) {
+		fwd_e = &config->fwd_entries[i];
+
+		if (odph_iplookup_table_put_value(config->fwd_tbl, &fwd_e->prefix, &fwd_e) < 0) {
+			ODPH_ERR("Error populating forwarding table\n");
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static inline odp_ipsec_sa_t *get_out_sa(odp_packet_t pkt)
+{
+	odph_udphdr_t *udp = odp_packet_l4_ptr(pkt, NULL);
+
+	return udp->dst_port ? spi_to_sa_map[DIR_OUT][odp_be_to_cpu_16(udp->dst_port)] : NULL;
+}
+
+static inline odp_bool_t do_ipsec_out(odp_packet_t pkt, const odp_ipsec_sa_t *sa)
+{
+	odp_ipsec_out_param_t param;
+
+	memset(&param, 0, sizeof(param));
+	param.num_sa = 1;
+	param.sa = sa;
+	/* IPsec in/out need to be identified somehow, so use user_ptr for this. */
+	odp_packet_user_ptr_set(pkt, sa);
+
+	return odp_ipsec_out_enq(&pkt, 1, &param) == 1;
+}
+
+static inline pkt_handle_result_t handle_udp(odp_packet_t pkt, stats_t *stats)
+{
+	odp_ipsec_sa_t *sa = get_out_sa(pkt);
+
+	if (sa == NULL)
+		return PKT_FWD;
+
+	if (odp_unlikely(!do_ipsec_out(pkt, sa))) {
+		++stats->ipsec_out_errs;
+		return PKT_NOK;
+	}
+
+	return PKT_OK;
+}
+
+static inline odp_ipsec_sa_t *get_in_sa(odp_packet_t pkt)
+{
+	odph_esphdr_t *esp = odp_packet_l4_ptr(pkt, NULL);
+	uint32_t spi = odp_be_to_cpu_32(esp->spi);
+
+	return spi <= UINT16_MAX ? spi_to_sa_map[DIR_IN][spi] : NULL;
+}
+
+static inline odp_bool_t do_ipsec_in(odp_packet_t pkt, const odp_ipsec_sa_t *sa)
+{
+	odp_ipsec_in_param_t param;
+
+	memset(&param, 0, sizeof(param));
+	param.num_sa = 1;
+	param.sa = sa;
+	odp_packet_user_ptr_set(pkt, NULL);
+
+	return odp_ipsec_in_enq(&pkt, 1, &param) == 1;
+}
+
+static inline pkt_handle_result_t handle_ipsec(odp_packet_t pkt, stats_t *stats)
+{
+	odp_ipsec_sa_t *sa = get_in_sa(pkt);
+
+	if (sa == NULL)
+		return PKT_FWD;
+
+	if (odp_unlikely(!do_ipsec_in(pkt, sa))) {
+		++stats->ipsec_in_errs;
+		return PKT_NOK;
+	}
+
+	return PKT_OK;
+}
+
+static inline fwd_entry_t *lookup(odp_packet_t pkt, odph_table_t fwd_tbl)
+{
+	odph_ethhdr_t *eth = odp_packet_l2_ptr(pkt, NULL);
```

Comment:
l2_ptr is not guaranteed to be valid after IPsec processing. L3 offset and l3_ptr are valid but not necessarily zero. You need to write the full Ethernet header, including ethertype, to the packet after first truncating or extending the packet head to make the L2 header the right size. It is a bit cumbersome but it is what the API requires.

There is also no guarantee that an IPsec-processed packet is contiguous in memory, so some checking is needed wherever pointers to packet data are used (also in ipsec input processing).

Reply:
Data accesses should be safer now in v3.